cdc: add builtin function to emulate json encoding

HonoreDB opened the pull request:
There's been a [longstanding debugging request](https://github.com/cockroachdb/cockroach/issues/75730) to be able to see how a row would get encoded by a changefeed with certain options. The main obstacle was removing the need for the builtins package to depend on any package implementing a builtin, which got easier thanks to the introduction of the builtinsregistry. This PR builds on that work to make everything that references builtin functions load-order independent. The first commit removes the load order effect on function OIDs, and the second changes everything that iterates over all builtins to instead register a hook to be called on each builtin as it gets registered. Finally, we add a crdb_internal function that's registered and implemented outside of the builtins package.

Release note (sql change): Added the crdb_internal.to_json_as_changefeed_with_flags function to help
debug json changefeeds. 

Release justification: Improves stability by ensuring that version changes don't alter OIDs and helping with debugging.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84509)
<!-- Reviewable:end -->

HonoreDB commented:
Sort of naturally composes with cdc expressions: `SELECT crdb_internal.to_json_as_changefeed_with_flags(projection.*, 'updated', 'diff') from [select address as place, 'dog' as hungry from users where id != 'ae147ae1-47ae-4800-8000-000000000022'] as projection limit 2;`

But we'd need very different function resolution logic to be able to use the cdc builtins there.

miretskiy commented on pkg/sql/sem/builtins/generator_builtins.go:
```diff
@@ -46,7 +46,7 @@ import (
 var _ eval.ValueGenerator = &arrayValueGenerator{}
 
-func initGeneratorBuiltins() {
+func init() {
```
This is an okay change; but it probably should be done in a separate PR if you feel
strongly about this cleanup.

miretskiy commented on pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go:
```diff
@@ -20,16 +20,40 @@ import "github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+// Subscription is a hook to be called once on all registered builtins.
+type Subscription func(name string, props *tree.FunctionProperties, overloads []tree.Overload)
+
+var subscriptions = make([]Subscription, 0)
```
making subscription w/ 0 len is the same as `var subscriptions []Subscription`

miretskiy commented on pkg/ccl/utilccl/builtins.go:
```diff
@@ -0,0 +1,214 @@
+// impl must be a function that returns a single type with a key in BuiltinReturnTypes,
+// plus optionally an error. This is a convenience method meant for generating SQL functions
+// that are tightly coupled to other behavior. It's generally preferable to create SQL builtins
+// via calling builtinsregistry.Register directly.
```
I honestly don't know about this -- seems like too much magic.

As your comment says that it is "preferable to create and register via builtinsregistry"... why not make it required? What does this function add that builtinregistry doesn't provide in a very explicit way?

miretskiy commented on pkg/ccl/changefeedccl/cdcevent/event.go:
```diff
@@ -530,3 +532,29 @@ func TestingGetFamilyIDFromKey(
+// MakeRowFromTuple converts a SQL datum produced by, for example, SELECT ROW(foo.*),
+// into the same kind of cdcevent.Row you'd get as a result of an insert, but without
+// the primary key.
+func MakeRowFromTuple(t *tree.DTuple) Row {
```
Very nice; this functions wants to be tested :)

miretskiy commented on pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go:
```diff
@@ -20,16 +20,40 @@ import "github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+// Subscriptions should be added at init() time, but are load-order independent:
+// if you add a subscription after a function is registered, it will immediately
+// be called on that function, while functions that are registered afterwards will
+// also trigger the hook.
```
having a bit of a hard time seeing why we need this.
Can you explain what is the purpose of this?
Does it make sense to pull it into separate commit?

miretskiy commented on pkg/ccl/changefeedccl/encoder_json.go:
```diff
@@ -195,3 +198,40 @@ func (e *jsonEncoder) EncodeResolvedTimestamp(
+}
+
+func init() {
+	utilccl.RegisterArgType(cdcevent.Row{}, types.AnyTuple, func(d tree.Datum) interface{} {
```
I just don't see how this is more readable than registering builtin where you
explicitly say that the function accepts a DTuple, and that you convert that tuple into a row, and then execute conversion.

miretskiy commented on pkg/sql/sem/tree/overload.go:
```diff
@@ -134,7 +134,7 @@ type Overload struct {
 
 	// OnTypeCheck is incremented every time this overload is type checked.
-	OnTypeCheck func()
+	OnTypeCheck *func()
```
I think this needs some explanation.. It's very infrequent that I see a function pointer...

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @shermanCRL)

___
*[`pkg/ccl/utilccl/builtins.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72q2hZcDy60yjp_zyy-r1-99:-N7GYikS24AII4a3ea9j:bj1vbao) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/utilccl/builtins.go#L99)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I honestly don't know about this -- seems like too much magic.

As your comment says that it is "preferable to create and register via builtinsregistry"... why not make it required? What does this function add that builtinregistry doesn't provide in a very explicit way?
</blockquote></details>

...yeah, I don't know if I can really defend this. My goal was to make it possible to add builtin functions without having to learn all the unusual idioms of SQL function definition code. Right now to define a changefeed-related function you need as an engineer to understand both changefeed code and SQL builtins code and you need to have a single file that depends on both `changefeedccl` and a bunch of `tree` objects. But once this file gets a few type mappings into it, you'll instead be able to write code that's normal for your package.

That said, I don't think "people should be able to write builtins without knowing what a tree.Overload is" is necessarily a good idea.
___
*[`pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01e-r1-51:-N7GWcEtEBdkIe_jPF95:btm0f9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go#L51)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

having a bit of a hard time seeing why we need this.
Can you explain what is the purpose of this?
Does it make sense to pull it into separate commit?
</blockquote></details>

The reason is that I don't know how to control load order without adding a dependency. There is probably a way? If I could guarantee that everything that needed to know about all builtins would load after everything that registered a builtin, I could just add a builtinsregistry.Freeze() function to make sure it stayed that way by panicking if you tried to register a builtin after calling a method like Iterate or AllBuiltinNames. But if we're going to register builtins in ccl packages, we can't have all_builtins take a dependency on every file that creates a builtin. The other other way to do it would be to have ccl builtins get registered no matter what and just throw an error in open source builds, but I feel like it's worth doing a little work to avoid that.
___
*[`pkg/sql/sem/tree/overload.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73USCj0-wD2HYn_0yL-r1-137:-N7GXxXGC8DNjxoBLAfK:b-7e16sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/tree/overload.go#L137)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think this needs some explanation.. It's very infrequent that I see a function pointer...
</blockquote></details>

I bet there was a better way to do this--I'll add a comment, but the idea was to be able to define this function in an AddSubscription hook when the overload is passed by value.


<!-- Sent from Reviewable.io -->

HonoreDB committed:
sql,ccl: allow builtins to be registered from anywhere

Previously, builtin functions needed to be created from
inside the builtins package. This lead to a giant package
with all the implementations, and also made it difficult
to register non-CCL builtins conditionally on the build
being non-oss. This commit makes builtin registration capable
of happening in any init() code by making everything that
needs to iterate over all builtins provide a hook instead
of forcing a particular build order.
This commit only leverages that to move sql telemetry logic
out of the builtins package. The following commit uses it
to create a CCL builtin function.

Release note: None

Release justification: Refactor necessary for CCL functions.

HonoreDB committed:
changefeedccl: add a sql builtin to emulate changefeed encoding

Informs #75730.

Release note (sql change): Added the crdb_internal.to_json_as_changefeed_with_flags function to help
debug json changefeeds.

Release justification: Additive change to help with debugging.

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @rafiss, and @shermanCRL)

___
*[`pkg/ccl/changefeedccl/encoder_json.go` line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01Z-r1-230:-N8AlWme97yMQXV9OBew:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/changefeedccl/encoder_json.go#L230)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I just don't see how this is more readable than registering builtin where you
explicitly say that the function accepts a DTuple, and that you convert that tuple into a row, and then execute conversion.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 539 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01X-r1-539:-N8AkdEtD6WfOwVnEb2J:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/changefeedccl/cdcevent/event.go#L539)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Very nice; this functions wants to be tested :)
</blockquote></details>

Added.
___
*[`pkg/ccl/utilccl/builtins.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72q2hZcDy60yjp_zyy-r1-99:-N8AkfzSAkrfYjt2Zs3e:b-t98xe2) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/utilccl/builtins.go#L99)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

...yeah, I don't know if I can really defend this. My goal was to make it possible to add builtin functions without having to learn all the unusual idioms of SQL function definition code. Right now to define a changefeed-related function you need as an engineer to understand both changefeed code and SQL builtins code and you need to have a single file that depends on both `changefeedccl` and a bunch of `tree` objects. But once this file gets a few type mappings into it, you'll instead be able to write code that's normal for your package.

That said, I don't think "people should be able to write builtins without knowing what a tree.Overload is" is necessarily a good idea.
</blockquote></details>

I have sighed wistfully and deleted all the magic.
___
*[`pkg/sql/crdb_internal.go` line 1578 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N8AkwdA1kCY4arDLIAm:-N8AkwdA1kCY4arDLIAn:b-tbhavb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f786aadcad04f3bfb7d7dce2bf138531d8d91f7c/pkg/sql/crdb_internal.go#L1578)):*
> ```Go
> 			}
> 
> 			if !hasModify && !hasView {
> ```

Reminder to self to make sure this diff isn't actually in the PR.
___
*[`pkg/sql/sem/builtins/all_builtins.go` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01c-r1-56:-N8Al3lb6nPPeqndp2s4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/all_builtins.go#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It would be nice to pull such refactoring into a separate commit (or even PR).
I'm pretty sure schema folks would have something to say about this.

I would prefer to have a better understanding (through nice commit message) why this change is needed (as I see it, it's because you want to add more things to those builtins, without adding the builtin implementation in this directory (and keeping it it under CCL).  But again, the mechanism for doing this is up for debate (or at least comments/approval from schema people).
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73d91R_2JrSvnCj-5K-r1-49:-N8Al6xO21X2MFcpA7c3:bkwyng0) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/generator_builtins.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is an okay change; but it probably should be done in a separate PR if you feel
strongly about this cleanup.
</blockquote></details>

I've pulled it into a separate commit. It sort of needs to be in this PR in order to prove that registration is build-order-independent.
___
*[`pkg/sql/sem/tree/overload.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73USCj0-wD2HYn_0yL-r1-137:-N8AlPwpEmqr8t9i9wTr:bkchjt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/tree/overload.go#L137)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I bet there was a better way to do this--I'll add a comment, but the idea was to be able to define this function in an AddSubscription hook when the overload is passed by value.
</blockquote></details>

(comment added)


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @rafiss, and @shermanCRL)

___
*[`pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01e-r1-26:-N8AmrR02cCIPS8oF0Rv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go#L26)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

making subscription w/ 0 len is the same as `var subscriptions []Subscription`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/cdcevent/event.go:
```diff
@@ -530,3 +532,29 @@ func TestingGetFamilyIDFromKey(
+				log.Warningf(context.Background(), "failed to set row value from tuple due to error %v", err)
+				_ = r.SetValueDatumAt(i, tree.DNull)
+			} else {
+				panic(err)
```
do we have to panic? Seems like returning error is fine?

miretskiy commented on pkg/ccl/changefeedccl/encoder_json.go:
```diff
@@ -195,3 +201,62 @@ func (e *jsonEncoder) EncodeResolvedTimestamp(
+	}
+
+	utilccl.RegisterCCLBuiltin("crdb_internal.to_json_as_changefeed_with_flags",
+		`Encodes a tuple the way a changefeed would output it if it were inserted as a row or emitted by a changefeed expression, and returns the raw bytes. 
```
`crdb_internal.cdc_encode_as_json`?
Or even `cdc_to_json` -- kinda similiar to `row_to_json`?

miretskiy reviewed (approved):
cdc stuff -- fine; the rest, somebody need to stamp it.

HonoreDB committed:
sql: make builtin function oids deterministic

Previously, oids were assigned to functions in the order
they were registered, so adding an alphabetically-earlier
function, overload, or file would change the oid. This was
mainly inconvenient for a few tests but also made it unsafe
to do things like serialize a function oid on a database
descriptor.

This commit hardcodes all builtin function oids to what
they were as of its immediate ancestor commit, and enforces
that newly-added overloads also get hardcoded.

The only reason this isn't a generator is that right now
running a generator is about equally onerous as adding one
line to fixed_oids.go. If that changes, this should be a generator.

Release justification: Low-impact refactor.
Release note: None

HonoreDB commented:
This is ready for a look on the SQL end, first two commits are builtins registration refactors.

rafiss reviewed (approved):
this lgtm, but @chengxiong-ruan may want to look at the first 2 commits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-:-NAk-gUE1Tjky6ubSJJf:bg0wkra)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @rafiss)

___
*[`pkg/sql/sem/builtins/all_builtins.go` line 32 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-NAjLaEk1hh4oYT3wMKy:-NAjLaEk1hh4oYT3wMKz:b-50pdm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bdf8236a508a808b660f045df72755d4a473afb/pkg/sql/sem/builtins/all_builtins.go#L32)):*
> ```Go
> }
> 
> func (o *orderedStrings) add(s string) {
> ```

this could use comments saying the differences between `add` and `insert`
___
*[`pkg/sql/sem/builtins/all_builtins.go` line 47 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-NAjLlmI492gyqxNUqxY:-NAjLlmI492gyqxNUqxZ:bjlgg30) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bdf8236a508a808b660f045df72755d4a473afb/pkg/sql/sem/builtins/all_builtins.go#L47)):*
> ```Go
> }
> 
> func (o *orderedStrings) insert(s string) {
> ```

if you call `insert` on a non-sorted list, should it become sorted first?


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/sem/builtins/fixed_oids.go:
```diff
@@ -0,0 +1,2041 @@
+	`voidsend(void: void) -> bytes`:                                                                                                                                                 1994,
+	`width_bucket(operand: decimal, b1: decimal, b2: decimal, count: int) -> int`:                                                                                                   811,
+	`width_bucket(operand: int, b1: int, b2: int, count: int) -> int`:                                                                                                               812,
+	`width_bucket(operand: anyelement, thresholds: anyelement[]) -> int`:                                                             
```
nit: just return `existing max id + 1`, what do you think?
I guess the effort here is trying get any unused id in between, and that'd be helpful if we removed any builtin functions. I'd prefer `existing max id + 1` because: (1) removing builtin function could be super rare. (2) reuse an id seems not desirable since we're trying to make OIDs deterministic here.

HonoreDB replied:
Good point, done.

chengxiong-ruan commented on pkg/sql/sem/builtins/all_builtins.go:
```diff
@@ -14,63 +14,88 @@ import (
 // names, sorted in alphabetical order. This can be used for a
 // deterministic walk through the Builtins map.
-var AllBuiltinNames []string
+func AllBuiltinNames() []string {
```
+1, I'm not sure why this function is preferred over a sorted string slice. Trying make the string slice private maybe? If so, I think we may just sort it once in the `init` function right?

HonoreDB replied:
This actually just sorts the slice once--further calls to .sort() will short-circuit. It's happening here rather than in an init() so that we don't depend on load order to fully populate the slices before they're sorted.

chengxiong-ruan replied:
I see, thanks, for some reason I missed the sortedStrings implementation above.

chengxiong-ruan reviewed (approved):
LGTM, but curious with 2 questions.

HonoreDB commented:
bors r=[miretskiy,rafiss,chengxiong-ruan]
